google/cloud/pubsub/internal/tracing_batch_sink.cc

```diff
@@ -0,0 +1,217 @@
+  // span with no links and each child spans will contain links.
+  if (message_spans.size() <= max_otel_links) {
+    batch_sink_spans.push_back(MakeParent(
+        MakeLinks(message_spans.begin(), message_spans.end()), message_spans));
+    return batch_sink_spans;
+  }
+  batch_sink_spans.push_back(MakeParent({{}}, message_spans));
+  auto batch_sink_parent = batch_sink_spans.front();
+
+  auto cut = [&message_spans, max_otel_links](auto i) {
+    auto const batch_size = static_cast<std::ptrdiff_t>(max_otel_links);
+    return std::next(
+        i, std::min(batch_size, std::distance(i, message_spans.end())));
+  };
+  int count = 0;
+  for (auto i = message_spans.begin(); i != message_spans.end(); i = cut(i)) {
+    // Generates child spans with links between [i, min(i + batch_size, end))
+    // such that each child span will have exactly batch_size elements or less.
+    batch_sink_spans.push_back(
+        MakeChild(batch_sink_parent, count++, MakeLinks(i, cut(i))));
+  }
+
+  return batch_sink_spans;
+}
+
+/**
+ * Records spans related to a batch messages across calls and
+ * callbacks in the `BatchingPublisherConnection`.
+ */
+class TracingBatchSink : public BatchSink {
+ public:
+  explicit TracingBatchSink(std::shared_ptr<BatchSink> child, Options opts)
+      : child_(std::move(child)), options_(std::move(opts)) {}
+
+  ~TracingBatchSink() override = default;
+
+  void AddMessage(pubsub::Message const& m) override {
+    auto active_span = opentelemetry::trace::GetSpan(
+        opentelemetry::context::RuntimeContext::GetCurrent());
+    active_span->AddEvent("gl-cpp.added_to_batch");
+    {
+      std::lock_guard<std::mutex> lk(mu_);
+      message_spans_.push_back(std::move(active_span));
+    }
+    child_->AddMessage(std::move(m));
+  }
+
+  future<StatusOr<google::pubsub::v1::PublishResponse>> AsyncPublish(
+      google::pubsub::v1::PublishRequest request) override {
+    decltype(message_spans_) message_spans;
+    {
+      std::lock_guard<std::mutex> lk(mu_);
+      message_spans.swap(message_spans_);
+    }
+
+    auto batch_sink_spans = MakeBatchSinkSpans(message_spans, options_);
+
+    // The first span in `batch_sink_spans` is the parent to the other spans in
+    // the vector.
+    internal::OTelScope scope(batch_sink_spans.front());
+    return child_->AsyncPublish(std::move(request))
+        .then([oc = opentelemetry::context::RuntimeContext::GetCurrent(),
+               spans = std::move(batch_sink_spans),
+               message_spans = std::move(message_spans)](auto f) mutable {
+          for (auto& span : message_spans) {
+            span->AddEvent("gl-cpp.publish_end");
+          }
+          for (auto& span : spans) {
+            internal::EndSpan(*span);
+          }
+          internal::DetachOTelContext(oc);
+          return f;
+        });
+  }
+
+  void ResumePublish(std::string const& ordering_key) override {
+    child_->ResumePublish(ordering_key);
+  };
```

Comment:
nit: we should have a super simple unit test for `ResumePublish`.

Reply:
Done